### ISSUE TITLE
Unable to fetch data from the API

### ISSUE THREAD

**Issue by gpapin** [consolibyte/quickbooks-php#166]
Hi,

Despite the exempla application saying I'm connected the code for any of those exemple return an null value.
I'm a bit lost as I don't have any error or exceptions, and the code is fairly simple.

I've already check that there is data in my sandbox company.
I've also check that the apis call are made toward the sandbox api.
I've checked that the consumer secret + key I'm using are for test ones.

The interesting things is when I run ```quickbooks-php-master/docs/partner_platform/example_app_ipp_v3/debugging.php```
I get this : 
```
The bill add failed, and here is why:ERROR: 3100: [message=ApplicationAuthorizationFailed; errorCode=003100; statusCode=403, ]

REQUEST: POST https://sandbox-quickbooks.api.intuit.com/v3/company/123145730608574/bill?requestid=0b042802-8ab5-b5b4-adfe-7e027486deb6&minorversion=6 HTTP/1.1
Content-Type: application/xml
Authorization: OAuth realm="", oauth_signature_method="HMAC-SHA1", oauth_signature="XXXXXXXXXXXXXXXXXX",  oauth_nonce="o3bF8", oauth_timestamp="1490548014", oauth_token="XXXXXXXXXXXXXXXXXX", oauth_consumer_key="XXXXXXXXXXXXXXXXXX", oauth_version="1.0"
Content-Length: 150

<Bill xmlns="http://schema.intuit.com/finance/v3">
	<DocNumber>abc123</DocNumber>
	<TxnDate>2014-07-12</TxnDate>
	<VendorRef>9</VendorRef>
</Bill>


RESPONSE: HTTP/1.1 403 Forbidden
Server: nginx
Date: Mon, 27 Mar 2017 17:25:31 GMT
Content-Type: text/xml
Content-Length: 391
Connection: keep-alive
Keep-Alive: timeout=5

<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<IntuitResponse time="2017-03-27T10:25:31.293-07:00" xmlns="http://schema.intuit.com/finance/v3">
    <Fault type="SERVICE">
        <Error code="3100">
            <Message>message=ApplicationAuthorizationFailed; errorCode=003100; statusCode=403</Message>
            <Detail></Detail>
        </Error>
    </Fault>
</IntuitResponse>
```

This seems to indicate that the connection can't be authorized but the [following code](https://github.com/consolibyte/quickbooks-php/blob/master/docs/partner_platform/example_app_ipp_v3/config.php#L79) tells me that I am connected, so I am well confused...

**Comment by consolibyte** [comment]
Disconnect. Reconnect. Try again and let us know if you have better luck.

**Comment by gpapin** [comment]
I did it without any more luck unfortunately. I have the exact same issue.

**Comment by consolibyte** [comment]
Please post code I can use to reproduce the issue.

**Comment by gpapin** [comment]
At this point the code is literally the one from this repo, I'm running what's inside this exemple : https://github.com/consolibyte/quickbooks-php/tree/master/docs/partner_platform/example_app_ipp_v3

The only thing I changed is the apptoken and the OAuth Consumer Key/Secret I've setup a new testing app to give you the credentials and then I tested it and IT WORKED however if I revert to my previous tokens it still doesn't work...  I don't know if it makes any difference but on the working app I've literally just changed it's name, as on the other hand the app that does not work, is accepting production as some production settings etc... Obviously it's not production ready but I wanted to preview the process - Does it stop it from working in the sandbox ?

**Comment by consolibyte** [comment]
I can't help you without seeing the code you're using. If they are sandbox credentials (as you've indicated they are) then there's no security risk here.

**Comment by gpapin** [comment]
Any luck with this ? To me it looks like you can't access the sandbox once you setup your app for production. That would be very unfortunate for ongoing development but that's the only explanation I have

**Comment by consolibyte** [comment]
You can use an app both in production and dev mode at once. 

Here are my thoughts based on what you've told me so far: 

* Please post your code, and clearly indicate what is a production token/OAuth cred vs. a sandbox token/OAuth cred. 
* This alone is NOT enough to get a working app. Please go follow the quick-start guide. 

> The only thing I changed is the apptoken and the OAuth Consumer Key/Secret

* If all you're doing is swapping production OAuth app token/consumer key/consumer secret for sandbox app token/consumer key/consumer secret then obviously things WILL NOT work since the OAuth tokens you have will be invalid after you swap things, and since you'll be in sandbox mode vs. production mode in the code itself.

**Comment by gpapin** [comment]
>Please post your code

What code are you talking about exactly ? because I already told you I'm using exactly what's in this repo... 

Let me try to explain again. I have an application in my quickbooks developer account Let's call it application A. That application has 2 sets of keys (developement and production). I then cloned this project and opened the config file of one of the example (link above) in that config file I replaced the tokens with the **developpement** credential for application A. I then went to the webpage of the exemple connected successfully but as soon as I am connected I can't get any information back from the api so the page *thinks* I am connected but none of the examples works.
I then disconnected everything and started a second test.


In my second test I created a new application on quickbooks developer platform. This application is only available for development at this stage, let's call is application B.
I went back to this repo that I have locally edited the same config file to change the tokens with the one offered by application B. I then went to the demo webpage, ran the exemple and everything seems to work just fine. I can log in, I can run the example and see the data being reflected on my sandbox company etc.

I never tried to access live data, I only mentioned the production in application 'A' because it's the only thing that is different in my 2 tests from my point of view.

Does that make any more sense now ?

**Comment by consolibyte** [comment]
> What code are you talking about exactly ?

Any files you changed, including, but not limited to, your `config.php` file. 

> in that config file I replaced the tokens with the developpement credential for application A.
> I then went to the webpage of the exemple connected successfully

This ^^^ is NOT enough to get a working application. 

Please follow the quick-start guide, which also has you replace several URLs, a database connection string, and an encryption key. 

> I went back to this repo that I have locally edited the same config file to change the tokens with the one offered by application B.

This ^^^ will likely cause you problems, because now you're trying to use OAuth access tokens (which are stored in the DATABASE, and NOT in the code) from App A, with App B. You should remove the tokens from the database, and reconnect to QuickBooks.

**Comment by gpapin** [comment]
Oh I see where the confusion comes from now...
I created a mysql account / database to match what is required by the database connection string in the config file so I didn't need to change it, I didn't bother changing the encryption key as it's a demo app anyways.

As for the url I changed it too... just forgot to mention it. But as I said the example is **working** with the token from application 'B' so surely the setup is correct. 

I also forgot to mention that I've purged the database between the tests.

If that's helpful I can create 2 repositories one for each set of keys I'm using and give you access to that ?
